File: stripe/src/main/java/com/stripe/android/view/AddPaymentMethodActivity.java
package com.stripe.android.view;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.os.IBinder;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.annotation.VisibleForTesting;
import android.view.ViewGroup;
import android.view.WindowManager;

import com.stripe.android.ApiResultCallback;
import com.stripe.android.CustomerSession;
import com.stripe.android.PaymentConfiguration;
import com.stripe.android.R;
import com.stripe.android.Stripe;
import com.stripe.android.StripeError;
import com.stripe.android.model.PaymentMethod;
import com.stripe.android.model.PaymentMethodCreateParams;

import java.lang.ref.WeakReference;
import java.util.Objects;

import static com.stripe.android.PaymentSession.TOKEN_PAYMENT_SESSION;

/**
 * Activity used to display a {@link AddPaymentMethodView} and receive the resulting
 * {@link PaymentMethod} in the <code>Activity#onActivityResult(int, int, Intent)</code> of the
 * launching Activity.
 *
 * <p>Should be started with {@link AddPaymentMethodActivityStarter}.</p>
 */
public class AddPaymentMethodActivity extends StripeActivity {

    public static final String TOKEN_ADD_PAYMENT_METHOD_ACTIVITY = "AddPaymentMethodActivity";

    public static final String EXTRA_NEW_PAYMENT_METHOD = "new_payment_method";

    @Nullable private AddPaymentMethodView mAddPaymentMethodView;
    @Nullable private Stripe mStripe;

    private PaymentMethod.Type mPaymentMethodType;
    private boolean mStartedFromPaymentSession;
    private boolean mShouldAttachToCustomer;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        final AddPaymentMethodActivityStarter.Args args =
                AddPaymentMethodActivityStarter.Args.create(getIntent());
        final PaymentConfiguration paymentConfiguration;
        if (args.paymentConfiguration != null) {
            paymentConfiguration = args.paymentConfiguration;
        } else {
            paymentConfiguration = PaymentConfiguration.getInstance(this);
        }
        mStripe = new Stripe(getApplicationContext(), paymentConfiguration.getPublishableKey());
        mPaymentMethodType = args.paymentMethodType;

        configureView(args);

        mShouldAttachToCustomer = mPaymentMethodType.isReusable && args.shouldAttachToCustomer;
        mStartedFromPaymentSession = args.isPaymentSessionActive;

        if (mShouldAttachToCustomer && args.shouldInitCustomerSessionTokens) {
            initCustomerSessionTokens();
        }
    }

    private void configureView(@NonNull AddPaymentMethodActivityStarter.Args args) {
        getViewStub().setLayoutResource(R.layout.add_payment_method_layout);
        final ViewGroup contentRoot = (ViewGroup) getViewStub().inflate();

        mAddPaymentMethodView = createPaymentMethodView(args);
        contentRoot.addView(mAddPaymentMethodView);

        setTitle(getTitleStringRes());

        if (mPaymentMethodType == PaymentMethod.Type.Card) {
            getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE);
        }
    }

    @StringRes
    private int getTitleStringRes() {
        switch (mPaymentMethodType) {
            case Card: {
                return R.string.title_add_a_card;
            }
            case Fpx: {
                return R.string.title_bank_account;
            }
            default: {
                throw new IllegalArgumentException(
                        "Unsupported Payment Method type: " + mPaymentMethodType.code);
            }
        }
    }

    @NonNull
    private AddPaymentMethodView createPaymentMethodView(
            @NonNull AddPaymentMethodActivityStarter.Args args) {
        switch (mPaymentMethodType) {
            case Card: {
                return AddPaymentMethodCardView.create(this, args.shouldRequirePostalCode);
            }
            case Fpx: {
                return AddPaymentMethodFpxView.create(this);
            }
            default: {
                throw new IllegalArgumentException(
                        "Unsupported Payment Method type: " + mPaymentMethodType.code);
            }
        }
    }

    @VisibleForTesting
    void initCustomerSessionTokens() {
        CustomerSession.getInstance()
                .addProductUsageTokenIfValid(TOKEN_ADD_PAYMENT_METHOD_ACTIVITY);
        if (mStartedFromPaymentSession) {
            CustomerSession.getInstance().addProductUsageTokenIfValid(TOKEN_PAYMENT_SESSION);
        }
    }

    @Override
    protected void onActionSave() {
        createPaymentMethod(Objects.requireNonNull(mStripe),
                Objects.requireNonNull(mAddPaymentMethodView).getCreateParams());
    }

    @VisibleForTesting
    void createPaymentMethod(@NonNull Stripe stripe, @Nullable PaymentMethodCreateParams params) {
        if (params == null) {
            return;
        }

        setCommunicatingProgress(true);
        stripe.createPaymentMethod(params,
                new PaymentMethodCallbackImpl(this, mShouldAttachToCustomer));
    }

    private void attachPaymentMethodToCustomer(@NonNull final PaymentMethod paymentMethod) {
        CustomerSession.getInstance()
                .attachPaymentMethod(Objects.requireNonNull(paymentMethod.id),
                        new PaymentMethodRetrievalListenerImpl(this));
    }

    private void finishWithPaymentMethod(@NonNull PaymentMethod paymentMethod) {
        setCommunicatingProgress(false);
        setResult(RESULT_OK, new Intent().putExtra(EXTRA_NEW_PAYMENT_METHOD, paymentMethod));
        finish();
    }

    @Nullable
    IBinder getWindowToken() {
        return getViewStub().getWindowToken();
    }

    @Override
    protected void setCommunicatingProgress(boolean communicating) {
        super.setCommunicatingProgress(communicating);
        if (mAddPaymentMethodView != null) {
            mAddPaymentMethodView.setCommunicatingProgress(communicating);
        }
    }

    private static final class PaymentMethodCallbackImpl
            extends ActivityPaymentMethodCallback<AddPaymentMethodActivity> {

        private final boolean mUpdatesCustomer;

        private PaymentMethodCallbackImpl(@NonNull AddPaymentMethodActivity activity,
                                          boolean updatesCustomer) {
            super(activity);
            mUpdatesCustomer = updatesCustomer;
        }

        @Override
        public void onError(@NonNull Exception error) {
            final AddPaymentMethodActivity activity = getActivity();
            if (activity == null) {
                return;
            }

            activity.setCommunicatingProgress(false);
            // This error is independent of the CustomerSession, so we have to surface it here.
            activity.showError(error.getLocalizedMessage());
        }

        @Override
        public void onSuccess(@NonNull PaymentMethod paymentMethod) {
            final AddPaymentMethodActivity activity = getActivity();
            if (activity == null) {
                return;
            }

            if (mUpdatesCustomer) {
                activity.attachPaymentMethodToCustomer(paymentMethod);
            } else {
                activity.finishWithPaymentMethod(paymentMethod);
            }
        }
    }

    private static final class PaymentMethodRetrievalListenerImpl extends
            CustomerSession.ActivityPaymentMethodRetrievalListener<AddPaymentMethodActivity> {
        private PaymentMethodRetrievalListenerImpl(@NonNull AddPaymentMethodActivity activity) {
            super(activity);
        }

        @Override
        public void onPaymentMethodRetrieved(@NonNull PaymentMethod paymentMethod) {
            final AddPaymentMethodActivity activity = getActivity();
            if (activity == null) {
                return;
            }

            activity.finishWithPaymentMethod(paymentMethod);
        }

        @Override
        public void onError(int errorCode, @NonNull String errorMessage,
                            @Nullable StripeError stripeError) {
            final AddPaymentMethodActivity activity = getActivity();
            if (activity == null) {
                return;
            }

            // No need to show this error, because it will be broadcast
            // from the CustomerSession
            activity.setCommunicatingProgress(false);
        }
    }

    /**
     * Abstract implementation of {@link ApiResultCallback} that holds a {@link WeakReference}
     * to an <code>Activity</code> object.
     */
    private abstract static class ActivityPaymentMethodCallback<A extends Activity>
            implements ApiResultCallback<PaymentMethod> {
        @NonNull private final WeakReference<A> mActivityRef;

        ActivityPaymentMethodCallback(@NonNull A activity) {
            mActivityRef = new WeakReference<>(activity);
        }

        @Nullable
        public A getActivity() {
            return mActivityRef.get();
        }
    }
}

File: stripe/src/main/java/com/stripe/android/view/AddPaymentMethodActivity.kt
package com.stripe.android.view

import android.app.Activity
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.text.method.LinkMovementMethod
import android.text.util.Linkify
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.annotation.StringRes
import androidx.core.text.util.LinkifyCompat
import androidx.core.view.ViewCompat
import androidx.lifecycle.ViewModelProvider
import com.stripe.android.CustomerSession
import com.stripe.android.PaymentConfiguration
import com.stripe.android.R
import com.stripe.android.Stripe
import com.stripe.android.databinding.AddPaymentMethodActivityBinding
import com.stripe.android.model.PaymentMethod
import com.stripe.android.model.PaymentMethodCreateParams

/**
 * Activity used to display a [AddPaymentMethodView] and receive the resulting
 * [PaymentMethod] in the `Activity#onActivityResult(int, int, Intent)` of the
 * launching Activity.
 *
 *
 * Should be started with [AddPaymentMethodActivityStarter].
 */
class AddPaymentMethodActivity : StripeActivity() {

    private val args: AddPaymentMethodActivityStarter.Args by lazy {
        AddPaymentMethodActivityStarter.Args.create(intent)
    }

    private val stripe: Stripe by lazy {
        val paymentConfiguration = args.paymentConfiguration
            ?: PaymentConfiguration.getInstance(this)
        Stripe(
            applicationContext,
            publishableKey = paymentConfiguration.publishableKey,
            stripeAccountId = paymentConfiguration.stripeAccountId
        )
    }

    private val paymentMethodType: PaymentMethod.Type by lazy {
        args.paymentMethodType
    }

    private val shouldAttachToCustomer: Boolean by lazy {
        paymentMethodType.isReusable && args.shouldAttachToCustomer
    }

    private val addPaymentMethodView: AddPaymentMethodView by lazy {
        createPaymentMethodView(args).also {
            it.id = R.id.stripe_add_payment_method_form
        }
    }

    private val viewModel: AddPaymentMethodViewModel by lazy {
        ViewModelProvider(
            this,
            AddPaymentMethodViewModel.Factory(
                stripe,
                args
            )
        )[AddPaymentMethodViewModel::class.java]
    }

    private val titleStringRes: Int
        @StringRes
        get() {
            return when (paymentMethodType) {
                PaymentMethod.Type.Card -> {
                    R.string.title_add_a_card
                }
                PaymentMethod.Type.Fpx -> {
                    R.string.title_bank_account
                }
                else -> {
                    throw IllegalArgumentException(
                        "Unsupported Payment Method type: ${paymentMethodType.code}"
                    )
                }
            }
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        configureView(args)
        setResult(
            Activity.RESULT_OK,
            Intent()
                .putExtras(
                    AddPaymentMethodActivityStarter.Result.Canceled
                        .toBundle()
                )
        )
    }

    override fun onResume() {
        super.onResume()
        addPaymentMethodView.requestFocus()
    }

    private fun configureView(args: AddPaymentMethodActivityStarter.Args) {
        args.windowFlags?.let {
            window.addFlags(it)
        }

        viewStub.layoutResource = R.layout.add_payment_method_activity
        val scrollView = viewStub.inflate() as ViewGroup
        val viewBinding = AddPaymentMethodActivityBinding.bind(scrollView)

        viewBinding.root.addView(addPaymentMethodView)
        createFooterView(viewBinding.root)?.let {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP_MR1) {
                addPaymentMethodView.accessibilityTraversalBefore = it.id
                it.accessibilityTraversalAfter = addPaymentMethodView.id
            }
            viewBinding.root.addView(it)
        }

        setTitle(titleStringRes)
    }

    private fun createPaymentMethodView(
        args: AddPaymentMethodActivityStarter.Args
    ): AddPaymentMethodView {
        return when (paymentMethodType) {
            PaymentMethod.Type.Card -> {
                AddPaymentMethodCardView(
                    context = this,
                    billingAddressFields = args.billingAddressFields
                )
            }
            PaymentMethod.Type.Fpx -> {
                AddPaymentMethodFpxView.create(this)
            }
            else -> {
                throw IllegalArgumentException(
                    "Unsupported Payment Method type: ${paymentMethodType.code}"
                )
            }
        }
    }

    private fun createFooterView(
        contentRoot: ViewGroup
    ): View? {
        return if (args.addPaymentMethodFooterLayoutId > 0) {
            val footerView = layoutInflater.inflate(
                args.addPaymentMethodFooterLayoutId,
                contentRoot,
                false
            )
            footerView.id = R.id.stripe_add_payment_method_footer
            if (footerView is TextView) {
                LinkifyCompat.addLinks(footerView, Linkify.ALL)
                ViewCompat.enableAccessibleClickableSpanSupport(footerView)
                footerView.movementMethod = LinkMovementMethod.getInstance()
            }
            footerView
        } else {
            null
        }
    }

    public override fun onActionSave() {
        createPaymentMethod(viewModel, addPaymentMethodView.createParams)
    }

    internal fun createPaymentMethod(
        viewModel: AddPaymentMethodViewModel,
        params: PaymentMethodCreateParams?
    ) {
        if (params == null) {
            return
        }

        isProgressBarVisible = true

        viewModel.createPaymentMethod(params).observe(
            this,
            { result ->
                result.fold(
                    onSuccess = {
                        if (shouldAttachToCustomer) {
                            attachPaymentMethodToCustomer(it)
                        } else {
                            finishWithPaymentMethod(it)
                        }
                    },
                    onFailure = {
                        isProgressBarVisible = false
                        showError(it.message.orEmpty())
                    }
                )
            }
        )
    }

    private fun attachPaymentMethodToCustomer(paymentMethod: PaymentMethod) {
        runCatching {
            CustomerSession.getInstance()
        }.fold(
            onSuccess = { customerSession ->
                viewModel.attachPaymentMethod(
                    customerSession,
                    paymentMethod
                ).observe(
                    this,
                    { result ->
                        result.fold(
                            onSuccess = ::finishWithPaymentMethod,
                            onFailure = {
                                isProgressBarVisible = false
                                showError(it.message.orEmpty())
                            }
                        )
                    }
                )
            },
            onFailure = {
                finishWithResult(AddPaymentMethodActivityStarter.Result.Failure(it))
            }
        )
    }

    private fun finishWithPaymentMethod(paymentMethod: PaymentMethod) {
        finishWithResult(AddPaymentMethodActivityStarter.Result.Success(paymentMethod))
    }

    private fun finishWithResult(result: AddPaymentMethodActivityStarter.Result) {
        isProgressBarVisible = false
        setResult(
            Activity.RESULT_OK,
            Intent().putExtras(result.toBundle())
        )
        finish()
    }

    override fun onProgressBarVisibilityChanged(visible: Boolean) {
        addPaymentMethodView.setCommunicatingProgress(visible)
    }

    internal companion object {
        internal const val PRODUCT_TOKEN: String = "AddPaymentMethodActivity"
    }
}
